akka-stream/src/main/scala/akka/stream/io/Framing.scala

```diff
@@ -0,0 +1,173 @@
+package akka.stream.io
+
+import java.nio.ByteOrder
+
+import akka.stream.scaladsl.Flow
+import akka.stream.stage.{ TerminationDirective, PushPullStage, SyncDirective, Context }
+import akka.util.{ ByteIterator, ByteStringBuilder, ByteString }
+
+import scala.annotation.tailrec
+
+object Framing {
+
+  def delimiter(delimiterBytes: ByteString, maximumFrameLength: Int): Flow[ByteString, ByteString, Unit] =
+    Flow[ByteString].transform(() ⇒ new DelimiterFramingStage(delimiterBytes, maximumFrameLength))
+      .named("delimiterFraming")
+
+  def lines(delimiter: String, maximumLineBytes: Int): Flow[ByteString, String, Unit] =
+    Framing.delimiter(ByteString(delimiter), maximumLineBytes).map(_.utf8String)
```

Comment:
Utf8 is a variable length encoding, so `utf8String` doesn't work on arbitrary chunks of ByteStrings. I had to implement proper stream UTF8-encoding/decoding so I guess it would make sense to move `akka.http.impl.engine.ws.Utf8Decoder` to streams and also provide a streaming interface for it which should be trivial. It currently doesn't have a streaming interface because for websocket I had to embed it inside of another stage.

The same is true for the other direction. An `Array[Char]` is a sequence of UTF-16 code units (called "character" in java for legacy reasons). UTF-16 is a variable length encoding itself to encode 32-bit Unicode code points. So, an arbitrarily chunked stream of `Array[Char]` cannot be naively converted to Unicode code points because code points could be split in half.